### DISCUSSION TITLE
Adjacent node ids from ts.samples() are not guaranteed to come from the same individual

### DISCUSSION THREAD

**Discussion by grahamgower** [tskit-dev/tskit#2047] [General]
This assumption often holds though. One can check the assumption with the following code:
```python
ploidy = 2
nodes = ts.samples(population)
individual = np.reshape(ts.tables.nodes.individual[nodes], (-1, ploidy))
assert np.all(individual[:, [0]] == individual[:, 1:])
```

Presumably this affects many tskit functions, albeit perhaps not obviously. Like `ts.genotype_matrix()`, where the adjacent columns are not guaranteed to be from the same individual.

**Comment by hyanwong** [comment]
Although you are assuming everything is the same ploidy there. It's not impossible (I guess) to only have e.g. one of the two genomes from a diploid sample. I suspect this might happen with some historical samples (aDNA is sometimes pseudohaploid, which is a horrible hack)

**Reply by grahamgower** [reply]
Yes, that's true. But I think the assertion will fail in that case (assuming the reshape succeeded because `len(nodes) % ploidy == 0`).

**Reply by hyanwong** [reply]
If the nth and mth samples are both haploid then even if all other nodes are from diploid individuals, and the nodes are adjacent, the assert will fail because of mismatching of node numbers between rows n and m (I think).

Better might be to use `np.unique()` with `return_inverse` somehow, I think?

**Reply by grahamgower** [reply]
Personally, I'm using this assertion on the secondary assumption that all individuals have the same ploidy. So I don't mind that the assert fails if there is mixed ploidy.

**Comment by hyanwong** [comment]
I think the following does it regardless of ploidy etc, and also accounts for the possibility that some intermediate nodes may not be associated with individuals:

```
a = ts.tables.nodes.individual[ts.samples()]
concat_adjacent = a[np.diff(a, prepend=np.nan).astype(bool)]  # Any runs of the same number get collapsed into one number
concat_adjacent = concat_adjacent[concat_adjacent != tskit.NULL]  # remove cases where nodes have no individual
assert len(np.unique(concat_adjacent)) == len(concat_adjacent)  # if still dups, individuals IDs cannot be adjacent
```
